File: Python/Component.py
import numpy as np
import gr_config as CONFIG



# function compare_image_region(image, sub_image)
#     assert(image:size()[2] == sub_image:size(2))
#     assert(image:size()[3] == sub_image:size(3))

#     local nom = 0
#     local denom = 0

#     for i=1, sub_image:size()[2] do
#         for j=1, sub_image:size()[3] do
#             nom = nom + bit.band(image[{1, i, j}], sub_image[{1, i, j}])
#             denom = denom + bit.bor(image[{1, i, j}], sub_image[{1, i, j}])
#         end
#     end

#     local ratio = nom/denom

#     return ratio
# end

def compare_image_region(image, sub_image):
    """
    """

    nom = 0
    denom = 0

    image_bw = threshold_image(image)
    sub_image_bw = threshold_image(sub_image)

    for i in range(sub_image_bw.shape[0]):
        for j in range(sub_image_bw.shape[1]):
            nom += image_bw[i, j] and sub_image_bw[i, j]
            denom += image_bw[i, j] or sub_image_bw[i, j]

    ratio = nom/float(denom)

    return ratio


class Component(object):
    """

    """

    def __init__(self, word, min_x, max_x):
        """
        """

        self.begin = min_x
        self.end = max_x
        self._image = word.crop_image(min_x, max_x)
        self._prototypes = []

    def height(self):
        return self._image.shape[0]

    def width(self):
        return self._image.shape[1]

    def find_prototype_region(self, prototype):
        """
        """

        if self.height() < prototype.height() or self.width() < prototype.width():
            prototype_image = threshold(prototype.copy_and_resize(self.width(), self.height()))
        else:
            prototype_image = threshold(prototype._image)

        ratios = np.zeros((self.height() - prototype_image.shape[0] + 1,
                          self.width() - prototype_image.shape[1] + 1))

        component_image = threshold(self._image)



        for i in range(ratios.shape[0]):
            for j in range(ratios.shape[1]):
                cropped_image = component_image[i:prototype_image.shape[0],
                                                j:prototype_image.shape[1]]

                ratios[i, j] = compare_image_region(cropped_image, prototype_image)


        return ratios

    def __repr__(self):
        return self.__str__()

    def __str__(self):
        """
        """

        if self._prototypes:
            return str(self._prototypes)
        else:
            return CONFIG.UNKNOWN_COMPONENT

File: Component.lua
local Component = {}

function Component:height ()
    return self._image:size(2)
end

function Component:width ()
    return self._image:size(3)
end

-- function Component:__tostring ()
--     return self._string
-- end

function Component:new (word, start, e)
    local component = {}

    setmetatable(component, self)
    self.__index = self

    component.s = start
    component.e = e

    -- Create image for component with size of e - start + 1
    --      start: begin of component
    --      e: end of component
    -- + 1 needs to be added because of the fencepost error
    component._image = word._image:narrow(3, start, e - start + 1)
    component._string = UNKNOWN_COMPONENT

    return component
end

function Component:find_prototype_region (prototype)
    local prototype_image

    if self:height() < prototype:height() or self:width() < prototype:width() then
        prototype_image = prototype:copy_and_resize(self:width(), self:height()):apply(threshold)
    else
        prototype_image = prototype._image:clone():apply(threshold)
    end

    local ratios = torch.Tensor(self:height() - prototype_image:size(2) + 1,
                                self:width() - prototype_image:size(3) + 1)

    local component_image = self._image:clone():apply(threshold)

    for i=1, ratios:size(1) do
        for j=1, ratios:size(2) do
            local cropped_image = component_image:narrow(2, i, prototype_image:size(2))
                                                 :narrow(3, j, prototype_image:size(3))

            ratios[{i, j}] = compare_image_region(cropped_image, prototype_image)
        end
    end

    -- return ratios
    -- Looks for the coordinate with the highest value and returns its
    -- index
    for i=1, ratios:size(1) do
        for j=1, ratios:size(2) do
            if ratios[i][j] == ratios:max() then
                return i, j, ratios:max()
            end
        end
    end
end

return Component